### PR TITLE
Move Group to JavaScript

### PR DESCRIPTION
**Pull request by Trivo25**
This PR aims to move the Group implementation from OCaml to JavaScript, reducing OCaml dependencies as best as possible. The new Group implementation covers the same API surface as the old one and produces the exact same constraints (important for compatibility with already existent projects)

This will allow us to easier maintain more of the SnarkyJS elements by relying less on OCaml as well as giving us more control over what we are doing. Additionally, in the future this will enable a `snarkyjs/light` version by relying on native JS arithmetic instead of requiring users to load a lot of bindings for simple calculations -- builds on top of https://github.com/o1-labs/snarkyjs/pull/902 and is motivated by the same reasons

https://github.com/o1-labs/snarkyjs-bindings/pull/25

closes https://github.com/o1-labs/snarkyjs/issues/924

### CODE REVIEW

File: src/examples/primitive_constraint_system.ts
Comment (Trivo25) [inline src/examples/primitive_constraint_system.ts:3]: We didn't really have any regression tests that tested `Group` and its constraints, I just hacked this mock contract-like thing together to integrate it into the contract regression tests

File: src/examples/primitive_constraint_system.ts
Comment (Trivo25) [inline src/examples/primitive_constraint_system.ts:35]: this one adds some "tests" for Group constraints

File: src/examples/vk_regression.ts
Comment (Trivo25) [inline src/examples/vk_regression.ts:12]: some minor changes to accommodate the Group constraint tests, see above

File: src/lib/group.ts
Comment (Trivo25) [inline src/lib/group.ts:33]: Using the overloaded constructor so that the new implementation is equivalent to the old OCaml one, which allowed both `(x, y)` as well as `{x, y}` as arguments. I am not using the `FieldLike` or `GroupLike` type from above so users arent confused by it and see "the actual type".

File: src/lib/group.ts
Comment (Trivo25) [inline src/lib/group.ts:9]: just to make the internal code a bit more readable

File: src/lib/group.ts
Comment (Trivo25) [inline src/lib/group.ts:61]: I briefly talked to Gregor about this; the constructor doesn't check if a (constant) element is on the curve - it only does so if its a witness. Adding this would add constraints that could break existing contracts. That could be something for a future PR

File: src/lib/group.ts
Comment (Trivo25) [inline src/lib/group.ts:219]: I am still trying to figure out why this isn't producing the same set of constraints:

I tried implementing it with `x1.equals(x2).and(y1.equals(y2))` - which is (at least to my knowledge) equaivalent to the OCaml implementation of 

`Boolean.all [ Impl.Field.equal x1 x2; Impl.Field.equal y1 y2 ]`
where `Boolean.all` just calls `b1 && b2` for tuples. it produces the same amount of rows/gates but with slightly different coefficients?

File: src/lib/group.ts
Comment (Trivo25) [inline src/lib/group.ts:136]: I would like to move this entirely to JavaScript, but wanted to establish a baseline first

File: src/lib/group.ts
Comment (Trivo25) [inline src/lib/group.ts:175]: similar to the comment above

File: src/lib/group.ts
Comment (Trivo25) [inline src/lib/group.ts:196]: The OCaml implementation only checked if the `x` coordinates are equal, allowing devs to prove `g1 === g2` even if `g1.y !== g2.y`

File: src/lib/group.ts
Comment (Trivo25) [inline src/lib/group.ts:49]: i feel like this is a bit hacky, if anyone has recommendations let me know :D

File: src/lib/group.ts
Comment (Trivo25) [inline src/lib/group.ts:136]: I will move that implementation to pure JS as well, just wanted to establish a baseline first. 
I will stack another PR with the detailed implementation on top of this one soon

File: src/lib/group.ts
Comment (Trivo25) [inline src/lib/group.ts:175]: same as above

File: src/lib/group.test.ts
Comment (mitschabaude) [inline src/lib/group.test.ts:16]: love it!! I didn't know there was a neat example group element. we should make this the generator

File: src/lib/group.test.ts
Comment (mitschabaude) [inline src/lib/group.test.ts:36]: nit: would be more readable to reuse `g` here

File: src/lib/group.ts
Comment (mitschabaude) [inline src/lib/group.ts:89]: would be good to make the example code valid (I guess this one throws because it's not a valid group element)

File: src/lib/group.ts
Comment (mitschabaude) [inline src/lib/group.ts:50]: I notice there's no way to pass in a `zero` / `infinity` element here. however, not all methods enforce this restriction in the same way. for example, you can to `g.add(g.neg())` and will end up with the group element `(1, 1)` which supposedly stands for `zero`

I think the whole class should pick one consistent way of treating `zero`. It should be consistent between in-snark and out of snark, and could be one of the following:
1. we never ever represent `zero` as a Group. methods that would result in `zero` will throw
2. we have one unique way to represent `zero`. something like `(0, 0)` would make sense.

it's ok if you can use `zero` in some but not all operations on group elements. however, if the constructor and `Group.check()` doesn't allow `zero`, then I'd argue no method should allow you to create `zero`. In order to preserve constraints, I think this might be what you have to do

File: src/lib/group.ts
Comment (mitschabaude) [inline src/lib/group.ts:62]: this helper seems a bit dangerous since it's ignoring (and thus, allowing) `infinity` without guaranteeing a canonical `(x, y)` representation of the zero element. I recommend one of two simple options:
- since this seems to be only used once, inline it
- make this method throw when is encounters `infinity: true`

File: src/lib/group.ts
Comment (mitschabaude) [inline src/lib/group.ts:78]: the name here is confusing! it should be called `#toProjective()`, right?

File: src/lib/group.ts
Comment (mitschabaude) [inline src/lib/group.ts:101]: see my other comments about consistent treatment of `zero`. This seems to accept `(0, y)` and treats that as zero element. at the same time, adding `g` to `g.neg()` (which should result in zero) returns `(1, 1)` IIUC

File: src/lib/group.ts
Comment (mitschabaude) [inline src/lib/group.ts:139]: TODO (not in this PR): when `Scalar` move merged, use `scalar.isConstant()`

File: src/lib/group.ts
Comment (mitschabaude) [inline src/lib/group.ts:145]: while I appreciate keeping the `Snarky` interface low-level and having logic like reversing the scalar bits on the JS side, it would be good if a doccomment on `Snarky.group.scale` documented this unexpected behaviour

File: src/lib/group.ts
Comment (mitschabaude) [inline src/lib/group.ts:185]: this lgtm! if it results in a constraint change, I'd also try to switch the two equals constraints (it's OCaml so stuff is sometimes happening in reverse order :D)

File: src/lib/group.ts
Comment (mitschabaude) [inline src/lib/group.ts:180]: nit: does this need a `Bool` constructor?

File: src/lib/group.ts
Comment (mitschabaude) [inline src/lib/group.ts:229]: IMO all these extra static methods are weird and they should either be deprecated or be present consistently on all classes, i.e. added to `Field` for example

File: src/snarky.d.ts
Comment (mitschabaude) [inline src/snarky.d.ts:188]: nit: can we rename this to `assertOnCurve()` to make it more self-documenting?

**Review by mitschabaude — COMMENTED** [review]
Awesome work, I like how this keeps the OCaml API slim.

A few comments, most notably regarding consistent treatment of the `zero` Group element. I haven't looked up myself how the in-snark OCaml methods treat zero, so I'll leave it to you to identify a consistent way of doing it (and implement + ideally document it)

File: src/lib/group.ts
Comment (Trivo25) [reply src/lib/group.ts:89]: ah yes, I did that before I found `(-1, 2)` :laughing:

File: src/lib/group.ts
Comment (Trivo25) [reply src/lib/group.ts:185]: Ohh that was actually it, it was checking it in reverse

File: src/lib/group.ts
Comment (Trivo25) [reply src/lib/group.ts:229]: I am also not a massive fan, I would rather deprecate them

File: src/lib/group.ts
Comment (Trivo25) [reply src/lib/group.ts:185]: Oh nope, this actually still behaves strangely :thinking:

File: src/lib/group.ts
Comment (Trivo25) [reply src/lib/group.ts:50]: That's a good catch - I'll have to take another look at this.
OCaml returns `(0, 0)` for constants, whereas `(12664759760331458874453076485325239921471337210849432813230171084403110838275, 19449452489080454700052938888178047022259553573804486106032048451047634501628)` for variables

**Comment by mitschabaude** [comment]
note: this PR has some conflicts with the Scalar move. I prepared a branch `refactor/group-and-scalar` (snarkyjs + bindings) which combines the two, that can be merged into whichever of the two PRs gets merged second

File: src/lib/group.ts
Comment (Trivo25) [reply src/lib/group.ts:50]: although allowing certain functions to handle the `zero` element would require less constraints, I feel like we are limiting certain low-level operations that may require a `zero` element for certain use case, wdyt?

File: src/lib/group.ts
Comment (bkase) [reply src/lib/group.ts:229]: We can always recreate them with `bind` anyway -- JS devs are used to this; so I agree we should deprecate

**Review by bkase — CHANGES_REQUESTED** [review]
This great!

There are a lot of subtle security gotchas in this sort of code -- as a side-effect of moving more of our logic to TypeScript, it becomes more likely that auditors will be productive reviewing this kind of code. Maybe we should be making a queue of std library code that we should get third-party audited and figuring out when to prioritize it.

File: src/examples/vk_regression.ts
Comment (Trivo25) [reply src/examples/vk_regression.ts:1]: it already is running in the CI! The PR gets blocked if a constrained mismatch is detected

File: src/lib/group.ts
Comment (Trivo25) [reply src/lib/group.ts:50]: Gregor and I had a quick huddle, we thought about leaving it as simple as possible but making it uniform for now. I am about to move the entire 'add` algorithm to JS as well to gain a better understanding, because right now, the OCaml implementation also returns something "weird" for the `zero element (see comment above), We also had the idea of using a `Safe` and `Unsafe` namespace, similar to `Bool`

**Review by mitschabaude — APPROVED** [review]

File: src/lib/group.ts
Comment (bkase) [reply src/lib/group.ts:44]: safe and unsafe is a great idea 👍 , but yeah can come after

**Review by bkase — APPROVED** [review]

**Pull request merged**
